### PR TITLE
Feature GADT support for class type parameters

### PR DESCRIPTION
**Pull request by Ang9876**
- Register GADT constraint class type parameters from outer contexts in typedDefdef.
- Remove a condition to support GADT approximation for class type parameters.
- Add test cases in `tests/pos/class-gadt`.
- Ignore test cases `tests/pos/i4345.scala` and `i5735.scala` (issue #11220 and #11221)

### CODE REVIEW

**Review by abgruszecki — COMMENTED** [review]
Hey Ang, the code looks good. Good catch with the nested classes. Could you rewrite the code to look up "outer" classes instead of accumulating class type parameters? We usually try to avoid storing information in context that we can otherwise get.

When class B is nested inside A, then the owner of B should be A. You'll need to recursively get owners which are classes.

**Comment by Ang9876** [comment]
> Hey Ang, the code looks good. Good catch with the nested classes. Could you rewrite the code to look up "outer" classes instead of accumulating class type parameters? We usually try to avoid storing information in context that we can otherwise get.
> 
> When class B is nested inside A, then the owner of B should be A. You'll need to recursively get owners which are classes.

Hi Alex, thanks a lot for the review. I just changed the code and used `outersInterator`.
But I have a question about `flags`. If I use
```scala
takeWhile(!_.owner.flags.is(Method))
```
instead of
```scala
takeWhile(!_.owner.is(Method))
```
Building dotty would fail with an error 
```terminal
[error] ## Exception when compiling 478 sources to /.../dotty-1/compiler/../out/bootstrap/scala3-compiler-bootstrapped/scala-3.0.0-RC1/classes
```
What is the difference between `sym.is()` and `sym.flags.is()`?

**Comment by abgruszecki** [comment]
@Ang9876 there seem to be some optimisations in `Symbol#is`, perhaps you've found a problem with those optimisations. Let's check after merging your PR if the problem is still there and if so, open an issue.

File: compiler/src/dotty/tools/dotc/typer/Typer.scala [hunk 1]
```diff
@@ -2030,6 +2030,15 @@ class Typer extends Namer
...
       case untpd.TypeDefs(tparams) => tparams
     }
+
+    // Register GADT constraint for class type parameters from outer to inner class definition. (Useful when nested classes exist.) But do not cross a function definition.
+    if sym.flags.is(Method) then
+      rhsCtx.setFreshGADTBounds
+      ctx.outer.outersIterator.takeWhile(!_.owner.is(Method))
+        .filter(ctx => ctx.owner.isClass && ctx.owner.typeParams.nonEmpty)
+        .toList.reverse
+        .map(ctx => rhsCtx.gadt.addToConstraint(ctx.owner.typeParams))
```
Comment (allanrenucci) [inline compiler/src/dotty/tools/dotc/typer/Typer.scala:2040]:
```suggestion
        .foreach(ctx => rhsCtx.gadt.addToConstraint(ctx.owner.typeParams))
```

Comment (Ang9876) [reply compiler/src/dotty/tools/dotc/typer/Typer.scala]: Thank you very much!

**Review by abgruszecki — APPROVED** [review]
Alright, this looks good. Thanks for your work, @Ang9876 !

**Pull request merged**